shared/src/ledger/eth_bridge/vp/mod.rs

```diff
@@ -0,0 +1,433 @@
+        let (key_a, key_b) = match extract_valid_keys_changed(keys_changed)? {
+            Some((key_a, key_b)) => (key_a, key_b),
+            None => return Ok(false),
+        };
+        let sender = match extract_valid_sender(&self.ctx, key_a, key_b)? {
+            Some(sender) => sender,
+            None => return Ok(false),
+        };
+        let authed = authorize::is_authorized(&self.ctx, tx_data, &sender)?;
+        Ok(authed)
+    }
+}
+
+/// If `keys_changed` represents a valid set of changed keys, return them,
+/// otherwise return `None`.
+fn extract_valid_keys_changed(
+    keys_changed: &BTreeSet<Key>,
+) -> Result<Option<(wrapped_erc20s::Key, wrapped_erc20s::Key)>, Error> {
+    // we aren't concerned with keys that changed outside of our account
+    let keys_changed: HashSet<_> = keys_changed
+        .iter()
+        .filter(|key| storage::is_eth_bridge_key(key))
+        .collect();
+    if keys_changed.is_empty() {
+        return Err(Error(eyre!(
+            "No keys changed under our account so this validity predicate \
+             shouldn't have been triggered"
+        )));
+    }
+    tracing::debug!(
+        relevant_keys.len = keys_changed.len(),
+        "Found keys changed under our account"
+    );
+
+    if keys_changed.len() != 2 {
+        tracing::debug!(
+            relevant_keys.len = keys_changed.len(),
+            "Rejecting transaction as only two keys should have changed"
+        );
+        return Ok(None);
+    }
+
+    let mut keys = HashSet::<_>::default();
+    for key in keys_changed.into_iter() {
+        let key = match wrapped_erc20s::Key::try_from(key) {
+            Ok(key) => {
+                // Until burning is implemented, we disallow changes to any
+                // supply keys via wasm transactions
+                if matches!(key.suffix, wrapped_erc20s::KeyType::Supply) {
+                    tracing::debug!(
+                        ?key,
+                        "Rejecting transaction as key is a supply key"
+                    );
+                    return Ok(None);
+                }
+                key
+            }
+            Err(error) => {
+                tracing::debug!(
+                    %key,
+                    ?error,
+                    "Rejecting transaction as key is not a wrapped ERC20 key"
+                );
+                return Ok(None);
+            }
+        };
+        keys.insert(key);
+    }
+
+    // We can .unwrap() here as we know for sure that this set has len=2
+    let (key_a, key_b) = keys.into_iter().collect_tuple().unwrap();
+    if key_a.asset != key_b.asset {
+        tracing::debug!(
+            ?key_a,
+            ?key_b,
+            "Rejecting transaction as keys are for different assets"
+        );
+        return Ok(None);
+    }
+    Ok(Some((key_a, key_b)))
+}
+
+/// Returns the `Address` which should be authorizing the balance change, or
+/// `None` if the balance change is invalid.
+fn extract_valid_sender(
```

Comment:
The name of this method is not very accurate. This mainly checks that balances were adjust correctly,

Reply:
Yeah this was not well named, have updated the name + docstring